test/ha/activator_test.go

```diff
@@ -0,0 +1,154 @@
+// +build e2e
+
+/*
+Copyright 2020 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ha
+
+import (
+	"log"
+	"sort"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"knative.dev/pkg/ptr"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/serving/pkg/apis/autoscaling"
+	revisionresourcenames "knative.dev/serving/pkg/reconciler/revision/resources/names"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/e2e"
+)
+
+const (
+	activatorDeploymentName = "activator"
+	activatorLabel          = "app=activator"
+	SLO                     = 0.99 // a minimum of requests are expected to fail when killing the activator until the backup takes over
+)
+
+// The Activator does not have leader election enabled.
+// The test ensures that stopping one of the activator pods doesn't affect user applications.
+// One service is probed during activator restarts and another service is used for testing
+// that we can scale from zero after activator restart.
+func TestActivatorHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	if err := waitForDeploymentScale(clients, activatorDeploymentName, haReplicas); err != nil {
+		t.Fatalf("Deployment %s not scaled to %d: %v", activatorDeploymentName, haReplicas, err)
+	}
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(test.ServingFlags.SystemNamespace).List(metav1.ListOptions{
+		LabelSelector: activatorLabel,
+	})
+	if err != nil {
+		t.Fatalf("Failed to get activator pods: %v", err)
+	}
+	// Sort the pods according to creation timestamp so that we can always kill the oldest one.
```

Comment:
Is there a reason for that? If so, add a comment to why.

Reply:
I was thinking about sticky sessions and making sure that we kill the in the order in which they start receiving traffic. But maybe that's not required.

Reply:
I don't think that matters much... 🤷‍♂ Especially for the first one we're killing, since they are all about the same age created during test initialization.